### ISSUE TITLE
Created a basic Azure Blob ImageResolver

### ISSUE THREAD

**Issue by jakenuts** [SixLabors/ImageSharp.Web#25]
Not sure of a better place to post this but here is a basic Azure Blob IImageResolver. 

https://gist.github.com/jakenuts/8e05fbd275e060775b6999bdf6d60c22

Thanks!

James

**Comment by JimBobSquarePants** [comment]
Ah this is fantastic. Something we definitely planned to do. 

I'll have a look at making a repo based on your work. 

Thanks! 👍

**Comment by JimBobSquarePants** [comment]
Work has started on this. 
https://github.com/SixLabors/ImageSharp.Web/tree/2d784e65f445ff493b28870de8c1097ee648e8d0/src/ImageSharp.Web.Providers.Azure

**Comment by izzet** [comment]
Could you please update us about the development of this provider?

**Comment by JimBobSquarePants** [comment]
Haven't had the opportunity to work on it in a few months

https://github.com/SixLabors/ImageSharp.Web/tree/js/azure-blob

**Comment by Pandorax100** [comment]
I noticed the azure-blob branch is a bit stale, and that there have been updates to master in the months since started this has been started; I've created a fork from the current repository and updated this azure provider with the current changes to master.

https://github.com/Pandorax100/ImageSharp.Web/tree/azure-blob

I know you have an open pull request already relating to this, but as it's been open for months with no updates, I was wondering what you think the best course of action is in regards to incorporating these changes?

**Comment by JimBobSquarePants** [comment]
To keep the current work in the open PR. Don't duplicate, it's poor form.

**Comment by JimBobSquarePants** [comment]
I've done a lot more work on the Azure resolver today and am nearly ready to ship it to the nightlies.

**Comment by Pandorax100** [comment]
Do you think it would be a good idea to have the Provider match on a configurable url prefix which is then removed from the path after matching?

See below to see what I mean:
https://github.com/Pandorax100/ImageSharp.Web/commit/274477f8ca177c6fe7d0abdca39bd23169ee076e

**Comment by JimBobSquarePants** [comment]
That's for your individual implementation to determine and is configurable at `Startup`.

Most people will use a single provider in their setup and can do so without any additional configuration. Our API is designed to allow the maximum flexibility for the consumer which is far more powerful than defining match strategies within the classes themselves.

For example. In the code below I register both the `PhysicalFileSystemProvider` and the `AzureBlobStorageImageProvider` with custom `Match` properties to allow them to coexist.

``` c#
IImageProvider PhysicalProviderFactory(IServiceProvider provider)
{
    var p = new PhysicalFileSystemProvider(
        provider.GetRequiredService<IHostingEnvironment>(),
        provider.GetRequiredService<FormatUtilities>())
    {
        Match = context =>
        {
            return !context.Request.Path.StartsWithSegments("/azure");
        }
    };

    return p;
}

IImageProvider AzureProviderFactory(IServiceProvider provider)
{
    var p = new AzureBlobStorageImageProvider(
        provider.GetRequiredService<IOptions<AzureBlobStorageImageProvider>>(),
        provider.GetRequiredService<FormatUtilities>())
    {
        Match = context =>
        {
            return context.Request.Path.StartsWithSegments("/azure");
        }
    };

    return p;
}

services.AddImageSharpCore()
    .SetRequestParser<QueryCollectionRequestParser>()
    .SetCache<PhysicalFileSystemCache>()
    .SetCacheHash<CacheHash>()
    .AddProvider(AzureProviderFactory)
    .AddProvider(PhysicalProviderFactory)
    .AddProcessor<ResizeWebProcessor>()
    .AddProcessor<FormatWebProcessor>()
    .AddProcessor<BackgroundColorWebProcessor>();
```

**Comment by JimBobSquarePants** [comment]
@Pandorax100 

I'm not happy about this. You know a plugin is in development and the API is in a state of flux.

https://www.nuget.org/packages/Pandorax.ImageSharp.Web.Azure/1.0.0-alpha2

**Comment by Pandorax100** [comment]
@JimBobSquarePants 

The intention was to remove functionality from that as it appears in this; not to maintain a separate library of azure providers. I just wanted to experiment with azure providers in a few personal projects I've been working on and a NuGet package appears to be the best way to include it in all of them.

Ideally I want to contribute to this repo, but as there's already an open pull request on the issue I wasn't aware that I was able to without being an approved contributor of that branch.

Other than the azure provider, the only thing that library provides is a default implementation of an azure blob cache (and a few convenience service collection extension methods).

Once the current pull request is merged (or should the cache implementation also be in that pull request?) I'd be more than happy to contribute a default azure blob cache to this repo - if it's something you're looking at having - then remove that package as it is no longer needed.

**Comment by JimBobSquarePants** [comment]
@Pandorax100 I'm more than happy for you to experiment and build your own providers, and even more happy to accept any help you can offer. It's the location of the packages you've deployed. It muddied the water for people searching for an Azure provider. There's much better places to save Nuget packages in this scenario including publishing them to your own file system.

That said, I can see you have unlisted the package so thank you very much. 

On the note of a cache. We can implement a separate PR and project for that. I want to keep everything decoupled and I've found when developing ImageProcessor.Web that a common scenario is to use a common provider across development environments but individual physical cache instances.

**Comment by JimBobSquarePants** [comment]
Ok. New provider is ready for review. Once merged this will be available via the MyGet repository. 

https://github.com/SixLabors/ImageSharp.Web/pull/49

@Pandorax100 You'll be happy to note I ended up following your match pattern for the provider. It was a great idea!

**Comment by Pandorax100** [comment]
Awesome, looks good!

I'm also very pleased with your changes to the ImageSharpCoreBuilderExtensions here https://github.com/SixLabors/ImageSharp.Web/commit/ab2a485c3fc168b1d871d9099a9b70bae5961fd8 as that's caused me issues in the past too

On the cache: Yeah, I can see why keeping everything decoupled would be preferable. However, seeing as the resolver for both the cache and the provider are going to be pretty much identical, what are your thoughts on a shared project between the cache and the provider so we don't have to duplicate the resolver?

**Comment by JimBobSquarePants** [comment]
> what are your thoughts on a shared project between the cache and the provider so we don't have to duplicate the resolver?

I can't make up my mind just now to be honest. I'm still trying to decide whether a blob cache is a good idea as the network traffic costs may outweigh the benefits.

**Comment by JimBobSquarePants** [comment]
I've merged the open PR and the package should now be available on our MyGet feed. Thanks all for your assistance and patience.
